prov/gni/src/gnix_av.c

```diff
@@ -36,13 +36,449 @@
 //
 #include <stdlib.h>
 #include <string.h>
+#include <inttypes.h>
 
 #include "gnix.h"
 #include "gnix_util.h"
 
+
+/*******************************************************************************
+ * Forward declarations of ops structures.
+ ******************************************************************************/
+static struct fi_ops_av gnix_av_ops;
+static struct fi_ops gnix_fi_av_ops;
+
+/*******************************************************************************
+ * Helper functions.
+ ******************************************************************************/
+/*
+ * TODO: Check for named AV creation and check RX CTX bits.
+ */
+static int gnix_verify_av_attr(struct fi_av_attr *attr)
+{
+	int ret = FI_SUCCESS;
+
+	switch (attr->type) {
+	case FI_AV_TABLE:
+	case FI_AV_MAP:
+		break;
+	default:
+		ret = -FI_EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+/*
+ * Check the capacity of the internal table used to represent FI_AV_TABLE type
+ * address vectors. Initially the table starts with a capacity and count of 0
+ * and the capacity increases by roughly double each time.
+ */
+static int gnix_check_capacity(struct gnix_fid_av *av, size_t count)
+{
+	struct addr_entry *addrs;
+	size_t capacity = av->capacity;
+
+	/*
+	 * av->count + count is the amount of used indices after adding the
+	 * count items.
+	 */
+	while (capacity < av->count + count) {
+		/*
+		 * Handle initial capacity of 0, by adding 1.
+		 */
+		capacity = capacity * 2 + 1;
+	}
+
+	/*
+	 * Don't need to grow the table.
+	 */
+	if (capacity == av->capacity) {
+		return FI_SUCCESS;
+	}
+
+	addrs = realloc(av->table, capacity * sizeof(*addrs));
+	if (!addrs) {
+		return -FI_ENOMEM;
+	}
+
+	/*
+	 * Update table and capacity to reflect new values.
+	 */
+	av->table = addrs;
+	av->capacity = capacity;
+
+	return FI_SUCCESS;
+}
+
+/*******************************************************************************
+ * FI_AV_TABLE specific implementations.
+ ******************************************************************************/
+static int table_insert(struct gnix_fid_av *int_av, const void *addr,
+			size_t count, fi_addr_t *fi_addr, uint64_t flags,
+			void *context)
+{
+	struct gnix_ep_name *temp;
+	int ret = count;
+	size_t index;
+	size_t i;
+
+	if (gnix_check_capacity(int_av, count)) {
```

Comment:
They use FI_SUCCESS in various places around libfabric in the same fashion so in the event that FI_SUCCESS does get redefined I'm sure we'd have time to handle it appropriately.